### DISCUSSION TITLE
Quaternion datatype

### DISCUSSION THREAD

**Discussion by josdejong** [josdejong/mathjs#2680] [Ideas]
There have been talks and a PR implementing a Quaternion datatype before, see #794:

> I have added a quaternion data type. I have tried to support as many functions as possible but have had issues getting the trigonometric functions to work with each other. Logs are also only supported to base e. Other than these exceptions quaternions are supported by the same functions that support complex numbers.
> 
> Quaternions can be created in any of the following ways:
> 
> ```js
> var a = math.quaternion(1,2,3,4);
> var b = math.quaternion({r:1, i:2, j:3, k:4});
> var c = math.quaternion('1 + 2i + 3j +4k');
> ```
> 
> There are other ways to create quaternions
> 
> Quaternion multiplication is noncommutative
> 
> ```js
> var a = math.quaternion(1,2,3,4);
> var b = math.quaternion(5,6,7,8);
> var c = math.multiply(a,b); // -60 + 12i +30j + 24k
> var d = math.multiply(b,a); // -60 + 20i + 14j + 32k
> math.unequal(c,d) // true
> ```
> 
> A final thing, this is the first open source project that I have contributed to so I apologize in advance if I've done something incorrectly.

This PR did stall. Let's discuss here if someone want to pick this up again.

**Comment by gwhitney** [comment]
Note that #2709 gives quaternions for free: A quaternion is just a complex number whose components are themselves complex numbers (so long as you introduce conjugation into the correct spots in complex multiplication, which is trivial to do and has no effect on the multiplication of ordinary complex numbers as conjugation does not change real numbers).

**Reply by josdejong** [reply]
ahh, good point. Thanks!
